wgsl/scanner.cc

```
@@ -0,0 +1,927 @@
+
+/// @param code_point the input code_point
+/// @return true if the code_point is part of the XIDStart unicode set
+bool is_xid_start(CodePoint code_point) {
+  // Fast path for ASCII.
+  if ((code_point >= 'a' && code_point <= 'z') ||
+      (code_point >= 'A' && code_point <= 'Z')) {
+    return true;
+  }
+  // With [a-zA-Z] handled, nothing less then the next sequence start can be
+  // XIDStart, so filter them all out. This catches most of the common symbols
+  // that are used in ASCII.
+  if (code_point < 0x000aa) {
+    return false;
+  }
+  return std::binary_search(kXIDStartRanges,
+                            kXIDStartRanges + kNumXIDStartRanges, code_point);
+}
+
+/// @param code_point the input code_point
+/// @return true if the code_point is part of the XIDContinue unicode set
+bool is_xid_continue(CodePoint code_point) {
+  // Short circuit ASCII. The binary search will find these last, but most
+  // of our current source is ASCII, so handle them quicker.
+  if ((code_point >= '0' && code_point <= '9') || code_point == '_') {
+    return true;
+  }
+  return is_xid_start(code_point) ||
+         std::binary_search(kXIDContinueRanges,
+                            kXIDContinueRanges + kNumXIDContinueRanges,
+                            code_point);
+}
+
+/// @return true if @p code_point is considered a whitespace
+bool is_space(CodePoint code_point) {
+  switch (code_point) {
+    case 0x0020:
+    case 0x0009:
+    case 0x000a:
+    case 0x000b:
+    case 0x000c:
+    case 0x000d:
+    case 0x0085:
+    case 0x200e:
+    case 0x200f:
+    case 0x2028:
+    case 0x2029:
+      return true;
+    default:
+      return false;
+  }
+}
+
+/// A fixed capacity, dynamic sized queue of bits (expressed as bools)
+template <size_t CAPACITY_IN_BITS>
+class BitQueue {
+ public:
+  /// @param index the index of the bit starting from the front
+  /// @return the bit value
+  auto operator[](size_t index) {
+    assert(index < count());
```

This should propagate an error out.